src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -83,36 +82,36 @@ public function connect()
             };
 
             $that->resolverPromises[Message::TYPE_AAAA] = $that->resolve(Message::TYPE_AAAA, $reject)->then($lookupResolve(Message::TYPE_AAAA));
-            $that->resolverPromises[Message::TYPE_A] = $that->resolve(Message::TYPE_A, $reject)->then(function (array $ips) use ($that, &$timer) {
+            $that->resolverPromises[Message::TYPE_A] = $that->resolve(Message::TYPE_A, $reject)->then(function (array $ips) use ($that) {
                 // happy path: IPv6 has resolved already (or could not resolve), continue with IPv4 addresses
                 if ($that->resolved[Message::TYPE_AAAA] === true || !$ips) {
                     return $ips;
                 }
 
                 // Otherwise delay processing IPv4 lookup until short timer passes or IPv6 resolves in the meantime
-                $deferred = new Promise\Deferred();
+                $deferred = new Promise\Deferred(function () use (&$ips) {
+                    // discard all IPv4 addresses if cancelled
+                    $ips = array();
+                });
                 $timer = $that->loop->addTimer($that::RESOLUTION_DELAY, function () use ($deferred, $ips) {
                     $deferred->resolve($ips);
                 });
 
-                $that->resolverPromises[Message::TYPE_AAAA]->then(function () use ($that, $timer, $deferred, $ips) {
+                $that->resolverPromises[Message::TYPE_AAAA]->then(function () use ($that, $timer, $deferred, &$ips) {
```

Comment:
This is the main source of the bug, isn't it? That one missing reference.

Reply:
The problem is actually quite nuanced, which is why it was so difficult to track down:

* The happy eyeballs algorithm may seem simple on the surface, but it is actually very detailed and complex because it establishes multiple connections concurrently.
* When a connection request is initiated, we simultaneously start DNS resolution for both IPv4 and IPv6, and then proceed to establish TCP/IP connections to the resulting addresses.
* If a connection request is cancelled, we reject the corresponding promise and ensure that any underlying resources are properly cleaned up.
* This specific problem occurs only when a connection request is cancelled within a 50ms timeframe when DNS for IPv4 has already resolved, while DNS for IPv6 is still pending.
* As part of the cancellation request, we cancel all pending connection requests (here: none) and then cancel all DNS queries (originally cancelling IPv6 before IPv4). Cancelling DNS for IPv6 has the effect of stopping the 50ms timer (as expected) and then adding the outstanding IPv4 addresses to the outstanding connection queue (not expected).
* As a consequence, we would start creating additional TCP/IP connections to any IPv4 addresses that would not otherwise be exposed to the outside. These connections would end up consuming resources until the server side closed them.
* The suggested solution ensures we cancel all DNS queries in reverse order (cancelling IPv4 before IPv6). Cancelling DNS for IPv4 first now has the effect of discarding any IPv4 addresses first before stopping the timer (as expected) and then adding an empty list of addresses to the outstanding connection queue, thus preventing any new connections.

Reply:
Yup that is rather nasty, well done :+1: .